### PR TITLE
feat(sdk):  Support new version template upgrade

### PR DESCRIPTION
**Pull request by kuangxiang20240501**
## Description

- support new mapping upgrading
  - control plane upgrade
  - stage data plane upgrade
  - production data plane upgrade
- new endpoint 
  - mapping template version list
  - api use case list 

## Related Issue
https://github.com/mycloudnexus/kraken/issues/30
https://github.com/mycloudnexus/kraken/issues/28

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)
- [ ] CI/CD or documentation update (changes to CI/CD pipeline or documentation)

## Checklist

- [ ] I ensured that the documentation is up to date
- [ ] I explained why this PR updates in detail with reasoning why it's required
- [ ] I would like a code coverage CI quality gate exception and have explained why

### CODE REVIEW

File: kraken-java-sdk/kraken-java-sdk-controller/src/main/java/com/consoleconnect/kraken/operator/controller/api/v3/TemplateUpgradeV3Controller.java [hunk 1]
```diff
@@ -27,7 +36,12 @@
 @Tag(name = "Template Upgrade V3", description = "Template Upgrade V3")
 @Slf4j
 public class TemplateUpgradeV3Controller {
+  public static final String UPGRADED = "Upgraded";
```
Comment (DaveXiong) [inline kraken-java-sdk/kraken-java-sdk-controller/src/main/java/com/consoleconnect/kraken/operator/controller/api/v3/TemplateUpgradeV3Controller.java:39]: Would you consider to define a enum for the status ?

File: kraken-java-sdk/kraken-java-sdk-controller/src/main/java/com/consoleconnect/kraken/operator/controller/api/v3/TemplateUpgradeV3Controller.java [hunk 2]
```diff
@@ -91,4 +105,79 @@ public HttpResponse<List<ComponentExpandDTO>> listApiUseCases(
...
+            .toList();
+    // the latest can upgrade
+    list.stream().findFirst().ifPresent(vo -> this.calculateStatus(vo, true));
+    list.stream().skip(1).forEach(vo -> this.calculateStatus(vo, false));
+    return HttpResponse.ok(
+        PagingHelper.toPageNoSubList(
+            list, assetDtoPaging.getPage(), assetDtoPaging.getSize(), assetDtoPaging.getTotal()));
+  }
+
+  protected void calculateStatus(TemplateUpgradeReleaseVO vo, boolean first) {
```
Comment (DaveXiong) [inline kraken-java-sdk/kraken-java-sdk-controller/src/main/java/com/consoleconnect/kraken/operator/controller/api/v3/TemplateUpgradeV3Controller.java:145]: better to move the business logic into the `service` layer

File: kraken-java-sdk/kraken-java-sdk-controller/src/main/java/com/consoleconnect/kraken/operator/controller/api/v3/TemplateUpgradeV3Controller.java [hunk 3]
```diff
@@ -91,4 +105,79 @@ public HttpResponse<List<ComponentExpandDTO>> listApiUseCases(
...
+        vo.setStatus(NOT_UPGRADED);
+        return;
+      }
+      vo.setStatus(UPGRADING);
+    } else {
+      vo.setStatus(DEPRECATED);
+    }
+  }
+
+  @Operation(summary = "stage environment upgrade check")
```
Comment (DaveXiong) [inline kraken-java-sdk/kraken-java-sdk-controller/src/main/java/com/consoleconnect/kraken/operator/controller/api/v3/TemplateUpgradeV3Controller.java:165]: what is `environment upgrade check` ?   this endpoint may cause confuse.

File: kraken-java-sdk/kraken-java-sdk-controller/src/main/java/com/consoleconnect/kraken/operator/controller/api/v3/TemplateUpgradeV3Controller.java [hunk 4]
```diff
@@ -91,4 +105,79 @@ public HttpResponse<List<ComponentExpandDTO>> listApiUseCases(
...
+      @PathVariable("productId") String productId,
+      @RequestParam(value = "templateUpgradeId", required = false) String templateUpgradeId,
+      @RequestParam String stageEnvId) {
+    return HttpResponse.ok(templateUpgradeService.stageCheck(templateUpgradeId, stageEnvId));
+  }
+
+  @Operation(summary = "product environment upgrade check")
+  @GetMapping("/production-upgrade-check")
+  public HttpResponse<List<String>> checkProductionUpgradeCondition(
+      @PathVariable("productId") String productId,
```
Comment (DaveXiong) [inline kraken-java-sdk/kraken-java-sdk-controller/src/main/java/com/consoleconnect/kraken/operator/controller/api/v3/TemplateUpgradeV3Controller.java:177]: the endpoint already indicates this is for `production`, but `productEnvId` is still required.

File: kraken-java-sdk/kraken-java-sdk-sync/src/main/java/com/consoleconnect/kraken/operator/sync/service/PullDeploymentService.java [hunk 5]
```diff
@@ -102,7 +102,12 @@ protected void ingestData(UnifiedAssetDto dto) {
 
   @Scheduled(cron = "${app.cron-job.pull-latest-release:-}")
   public void scheduledCheckLatestProductRelease() {
+    for (int i = 0; i < 5; i++) {
```
Comment (DaveXiong) [inline kraken-java-sdk/kraken-java-sdk-sync/src/main/java/com/consoleconnect/kraken/operator/sync/service/PullDeploymentService.java:105]: Can I know why there is a loop here ?  and why is it  `5` ?

**Review by DaveXiong — COMMENTED** [review]
comments added

File: kraken-java-sdk/kraken-java-sdk-controller/src/main/java/com/consoleconnect/kraken/operator/controller/api/v3/TemplateUpgradeV3Controller.java (hunk 4 above)
Comment (kuangxiang20240501) [reply kraken-java-sdk/kraken-java-sdk-controller/src/main/java/com/consoleconnect/kraken/operator/controller/api/v3/TemplateUpgradeV3Controller.java:177]: rename parameter productEnvId  to envId

File: kraken-java-sdk/kraken-java-sdk-controller/src/main/java/com/consoleconnect/kraken/operator/controller/api/v3/TemplateUpgradeV3Controller.java (hunk 3 above)
Comment (kuangxiang20240501) [reply kraken-java-sdk/kraken-java-sdk-controller/src/main/java/com/consoleconnect/kraken/operator/controller/api/v3/TemplateUpgradeV3Controller.java:165]: During the template upgrade, before the user clicks stage upgrade, there is a pre-check(e.g. whether mapping is completed) and the result needs to be displayed in the UI.

File: kraken-java-sdk/kraken-java-sdk-sync/src/main/java/com/consoleconnect/kraken/operator/sync/service/PullDeploymentService.java (hunk 5 above)
Comment (kuangxiang20240501) [reply kraken-java-sdk/kraken-java-sdk-sync/src/main/java/com/consoleconnect/kraken/operator/sync/service/PullDeploymentService.java:105]: The task is triggered every 30 seconds, and only one deployment can be pulled at a time, resulting in a long upgrade time.

Comment (DaveXiong) [reply kraken-java-sdk/kraken-java-sdk-sync/src/main/java/com/consoleconnect/kraken/operator/sync/service/PullDeploymentService.java:105]: for the above logic,  I can't fully understand it.   I would like other team members can review it based on the feature requirement, from my perspective, there maybe a issue here, 
1.  the scheduled can be customized via app.cron-job.pull-latest-release, it maybe shorter than 30s or can be longer than 30s
2. `5` is just a idea value based on the current test, if there are some factors changed,  `5` may need to be changed as well.

Comment (kuangxiang20240501) [reply kraken-java-sdk/kraken-java-sdk-sync/src/main/java/com/consoleconnect/kraken/operator/sync/service/PullDeploymentService.java:105]: For example, a template upgrade contains 16 use cases, each of which generates a separate deployment. Therefore, the data plane actually needs 16 API calls, each with an interval of 30 seconds, which takes about 5 minutes. This time is too long, so it is allowed to loop five times in one job to reduce the upgrade time.

File: kraken-java-sdk/kraken-java-sdk-controller/src/main/java/com/consoleconnect/kraken/operator/controller/api/v3/TemplateUpgradeV3Controller.java [hunk 6]
```diff
@@ -91,4 +105,79 @@ public HttpResponse<List<ComponentExpandDTO>> listApiUseCases(
...
+    if (first) {
+      if (CollectionUtils.isEmpty(vo.getDeployments())) {
+        vo.setStatus(NOT_UPGRADED);
+        return;
+      }
+      vo.setStatus(UPGRADING);
+    } else {
+      vo.setStatus(DEPRECATED);
+    }
+  }
```
Comment (xuelianhan007) [inline kraken-java-sdk/kraken-java-sdk-controller/src/main/java/com/consoleconnect/kraken/operator/controller/api/v3/TemplateUpgradeV3Controller.java]: Lines from 154 to 163 can be simplified by the following clause:
`vo.setStatus(first ? (CollectionUtils.isEmpty(vo.getDeployments()) ? NOT_UPGRADED : UPGRADING) : DEPRECATED);
`

Comment (kuangxiang20240501) [reply kraken-java-sdk/kraken-java-sdk-controller/src/main/java/com/consoleconnect/kraken/operator/controller/api/v3/TemplateUpgradeV3Controller.java]: > Lines from 154 to 163 can be simplified by the following clause: `vo.setStatus(first ? (CollectionUtils.isEmpty(vo.getDeployments()) ? NOT_UPGRADED : UPGRADING) : DEPRECATED); `
It seems that the code looks compact, but not easy to read.

File: kraken-java-sdk/kraken-java-sdk-sync/src/main/java/com/consoleconnect/kraken/operator/sync/service/PullDeploymentService.java (hunk 5 above)
Comment (zhiyaoPCCW) [reply kraken-java-sdk/kraken-java-sdk-sync/src/main/java/com/consoleconnect/kraken/operator/sync/service/PullDeploymentService.java:105]: mabye could consider to use an expression to caculate the times, for example: 
batchSize = (totalApiUsecases * 30 * perUserCaseCost) / expectCompleteTime

Configure the variables outside the yaml file to make this readable.

File: kraken-java-sdk/kraken-java-sdk-core/src/main/java/com/consoleconnect/kraken/operator/core/service/CompatibilityCheckService.java [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+  public boolean check(String appVersion, String productVersion) {
+    // read compatible asset from db
+    List<UnifiedAssetDto> assetDtos =
+        unifiedAssetService.findByKind(AssetKindEnum.PRODUCT_COMPATIBILITY.getKind());
+    if (CollectionUtils.isEmpty(assetDtos)) {
+      return true;
+    }
+    return StringUtils.isNoneEmpty(appVersion, productVersion);
+    // check whether match
```
[the remark this answers is not in the document]
Comment (kuangxiang20240501) [reply kraken-java-sdk/kraken-java-sdk-core/src/main/java/com/consoleconnect/kraken/operator/core/service/CompatibilityCheckService.java:24]: yeah,need to implement it later

File: kraken-java-sdk/kraken-java-sdk-controller/src/main/java/com/consoleconnect/kraken/operator/controller/service/SystemInfoService.java [hunk 8]
```diff
@@ -63,18 +63,18 @@ public void initialize() {
...
                 null,
                 null,
                 PageRequest.of(0, 1, Sort.Direction.DESC, AssetsConstants.FIELD_CREATE_AT),
                 null)
             .getData()
             .stream()
             .findFirst()
-            .map(t -> t.getMetadata().getLabels().get(LabelConstants.LABEL_PRODUCT_VERSION))
-            .map(t -> t.replaceFirst("[V|v]", ""))
+            .map(t -> t.getMetadata().getLabels().get(LabelConstants.LABEL_APP_TEMPLATE_UPGRADE_ID))
```
[the remark this answers is not in the document]
Comment (kuangxiang20240501) [reply kraken-java-sdk/kraken-java-sdk-controller/src/main/java/com/consoleconnect/kraken/operator/controller/service/SystemInfoService.java:74]: The labels default value is a empty map

File: kraken-java-sdk/kraken-java-sdk-controller/src/main/java/com/consoleconnect/kraken/operator/controller/handler/ClientHeartbeatEventHandler.java [hunk 9]
```diff
@@ -83,4 +88,18 @@ public Optional<EnvironmentEntity> queryEnvName(String envId) {
     }
     return Optional.empty();
   }
+
+  private void updateAppVersion(String envId, String appVersion) {
+    environmentRepository
+        .findById(UUID.fromString(envId))
+        .ifPresent(
+            env -> {
+              if (env.getName().equalsIgnoreCase(EnvNameEnum.STAGE.name())) {
```
[the remark this answers is not in the document]
Comment (kuangxiang20240501) [reply kraken-java-sdk/kraken-java-sdk-controller/src/main/java/com/consoleconnect/kraken/operator/controller/handler/ClientHeartbeatEventHandler.java:97]: In fact, after the pre-check, the env.name wont be null ,but it,s a good idea to switch the compare

**Review by xuelianhan007 — APPROVED** [review]
LGTM

**Review by zhiyaoPCCW — APPROVED** [review]
LGTM

**Pull request merged**